CorsixTH/Lua/endconditions.lua

```diff
@@ -0,0 +1,254 @@
+  {name = "value",            icon = 15, formats = 2}, -- Hospital value
+}
+
+-- This table supplies a function when a criteria cannot be measured
+--  directly from a hospital attribute of the same name.
+local get_custom_criteria = {
+  balance = function(hospital) return hospital.balance - hospital.loan end,
+}
+
+class "EndConditions"
+---@type EndConditions
+local EndConditions = _G["EndConditions"]
+
+--! Load conditions (goals) to win and lose from the map,
+-- and store them in world.win_goals and world.lose_goals.
+-- Put the highest number of group of conditions in world.highest_group.
+-- Determine winning and losing conditions in their groups.
+-- These groups are often incomplete, possibly empty
+-- ie more groups of conditions that lead to loss than win.
+--!param level_config (table) The map's level_config, containing criteria.
+--!param level_number (integer) The number of the map, if playing a campaign.
+function EndConditions:EndConditions(level_config, level_number)
+  local start, win_goals, lose_goals, highest_group, town = {}, {}, {}, 1
+  if level_config.towns then
+    if level_number and level_config.towns[level_number] then
+      town = level_config.towns[level_number]
+    else
+      town = level_config.town
+    end
+    start.balance = town.StartCash or 0
+    start.reputation = town.StartRep or 500
+  end
+
+  local win = level_config.win_criteria
+  if win then
+    for _, values in pairs(win) do
+      if values.Criteria ~= 0 then
+        local crit_name = local_criteria_variable[values.Criteria].name
+        if not win_goals[values.Group] then win_goals[values.Group] = {} end
+        win_goals[values.Group][crit_name] = {
+          name = crit_name,
+          win_value = values.Value,
+          boundary = values.Bound,
+          criterion = values.Criteria,
+          max_min_win = values.MaxMin,
+          icon = local_criteria_variable[values.Criteria].icon,
+          formats = local_criteria_variable[values.Criteria].formats,
+          start = start[crit_name] or 0,
+        }
+        if values.Group > highest_group then highest_group = values.Group end
+      end
+    end
+  end
+  local lose = level_config.lose_criteria
+  if lose then
+    for _, values in pairs(lose) do
+      if values.Criteria ~= 0 then
+        local crit_name = local_criteria_variable[values.Criteria].name
+        if not lose_goals[values.Group] then lose_goals[values.Group] = {} end
+        lose_goals[values.Group][crit_name] = {
+          name = crit_name,
+          lose_value = values.Value,
+          boundary = values.Bound,
+          criterion = values.Criteria,
+          max_min_lose = values.MaxMin,
+          icon = local_criteria_variable[values.Criteria].icon,
+          formats = local_criteria_variable[values.Criteria].formats,
+          start = start[crit_name] or 0,
+        }
+        if values.Group > highest_group then highest_group = values.Group end
+      end
+    end
+  end
```

Comment:
Looks like a copy of the win-criteria code just above it, except for the outer names `lose`, `lose_criteria`, and `lose_goals`.

If so, refactor this function to a private function and pass in the variables I just listed to the function, so you avoid copy/pasting the entire body.